Multiple selection

boscoa opened the pull request:
Je sais, j'ai fait un downcast. Je n'ai pas trouvé d'autres alternatives for now. Je suis ouvert aux commentaires.

boscoa committed:
Added composite pattern to select multiple shapes.

boscoa committed:
Added ungroup action. Moved vertices for loop inside selection box.

Ungroup button added to attributes panel. This will trigger an ActionEvent that clears the selectionBox and remove it from drawnShapes.

pspino commented on src/main/java/apaintus/controllers/AttributeController.java:
```diff
@@ -43,12 +44,15 @@ public void injectParentController(Controller controller) {
         createValueFactory();
         fillSpinnerMap();
+
+        ungroup.setVisible(false);
```
est-ce que cela fait juste grisonner le boutton ou cela le remove completement?

boscoa replied:
Disparaître complètement

pspino replied:
serait-il pas mieux de just le griser ? est-ce que cela affect le resizing de la toolbar ou attribute bar?

boscoa replied:
Il est à la fin et rien n'est affecté, de ce que j'ai pu tester jusqu'à maintenant
Personnellement, je n'aimerai pas avoir un bouton existant en gris que je peux utiliser qu'à certain moment.

pspino commented on src/main/java/apaintus/controllers/AttributeController.java:
```diff
@@ -83,15 +87,16 @@ public double getAttributeValue(Attribute attribute) {
         for (Map.Entry<Attribute, SpinnerValueFactory<Double>> entry : spinners.entrySet()) {
             entry.getValue().setValue(0.0);
         }
+        ungroup.setVisible(false);
```
same as previous

pspino commented on src/main/java/apaintus/controllers/CanvasController.java:
```diff
@@ -53,45 +57,57 @@ public void initialize() {
+            if (activeShape != null) {
+                activeShape.setSelected(false);
+                this.attributeController.resetAttributes();
+                redrawCanvas();
```
i get the redraw is to remove the bounding box, but is this the good way to do so ? imagine if we have 189273409871239487 shapes, it will lag juste to unselect the shape and remove the Bbox

pgarneau replied:
T'as raison @pspino , mais le probleme est comment le canvas prend du data. Il fait juste empiler des nouvelles couches de drawLayer a chaque saveState(), donc pour enlever quelque chose dans le canvas de base, il faut le redessiner au complet sans cette chose. Possiblement, on pourrait avoir un systeme de layers qui forment un "canvas" ensemble et simplement redraw la layer concernee, mais c'est pas implemente

pspino commented on src/main/java/apaintus/controllers/CanvasController.java:
```diff
@@ -160,7 +176,40 @@ private void saveDrawLayer() {
+            attributeController.resetAttributes();
+        }
+
+        canvasService.clear(drawLayer.getGraphicsContext2D());
```
save selection box shouldn't have to redraw the canvas??

boscoa replied:
En fait, puisque la selection box est resize pour fit les shapes à l'intérieur, il faut provoquer un redraw. Sinon, si on apply comme dans saveState, la bounding box de la selection box sera de la taille du user

pspino commented on src/main/java/apaintus/controllers/ToolBarController.java:
```diff
@@ -124,7 +124,7 @@ public ActiveTool getActiveTool() {
 
     private void setColorPicker(ColorPicker colorPicker, String colorValue) {
-        if(!colorValue.isEmpty()) {
+        if(colorValue != null &&  !colorValue.isEmpty()) {
```
is the color value is null it should throw an error since it is "not possible" to have a null color

pgarneau replied:
Pourquoi il y a une possibilite d'avoir null @boscoa ?

pspino replied:
je sais pas pourquoi le downvote garni, clairement, bosco a mis ce check parce que cela arrivait. Si jamais cela arrive, cela devrait throw une erreur et avertir que cela n'est pas possible, non? En plus de fixer la raison pourquoi cela arrive en ce moment aussi

boscoa replied:
Je n'ai jamais eu une String null dans Java jusqu'à maintenant, mais la SelectionBox provoquait un null sur la color quand elle était sélectionnée

pgarneau replied:
S'il le met pas, il va pogner un NullPointerException donc voila ton erreur. C'est plus important de verifier pourquoi null pour se retrouver a cet endroit la. Possiblement utiliser un Optional<> si necessaire.

boscoa replied:
J'ai trouvé la raison du null. Le truc, c'est dans la factory, je ne build pas avec le strokeColor sur la selectionBox car il s'agirait d'une information non pertinente et son parent Shape set à null sa valeur. En utilisant un Optional, faudrait faire une condition pour vérifier qu'il est présent de toute façon, non?

pgarneau replied:
Ouin... On a maybe un refactor a faire sur les objets qui peuvent etre dessines. Si on commence a avoir plus que des "shape", faudrait reviser comment on fait

boscoa replied:
Sinon, on pourrait ajouter dans les constructeurs des shapes des conditions de null... Du style : this.width = (shapeAttributes.getWidth() == null) ? 0 : shapeAttributes.getWidth();
À ce moment, on s'assure qu'aucune shape n'aura un élément set à null.

pspino commented on src/main/java/apaintus/services/draw/selection_box/SelectionBoxDrawService.java:
```diff
@@ -0,0 +1,19 @@
+package apaintus.services.draw.selection_box;
```
why should the drag_select have its own service when we have the drawService ?

pgarneau replied:
Parce que il y a une implementation unique a chaque truc qu'on dessine. Voir strategy pattern

pspino replied:
then on a bleeder un peu pas mal partout alors dans le code si on suis cette logique

boscoa replied:
Bleeder où?
Et j'en avais besoin d'un qui ne faisait rien puisque je ne voulais que print la bounding box.

pgarneau replied:
je ne comprend pas ton commentaire

pspino replied:
en ce moment drawService est appeler pour draw des shapes, le BoundingBox(qui est un shape, mais bon) et le snapgrid. 3 choses differentes l'un de lautre. Le SelectionBoxDrawService servirais a adder un Bounding box par desuss la selection du dragSelect ce qui m'apparait comme etant un service qu'offre deja le drawService pour les bounding box, est-ce que je me trompe?

boscoa replied:
J'ai voulu utiliser que le drawService du boundingBox pour la selectionBox, mais le canvasService appelle le drawService de la bounding box de la selected shape. Donc j'aurais introduit une redondance d'appel inutile à cet endroit, donc j'ai défini un drawService pour la Selection Box qui s'assure que rien n'est appelé sauf le drawService de la bounding box par le canvasService.

pspino replied:
ok

pgarneau replied:
@pspino DrawService c'est une classe abstraite tu peux pas l'instantier pis l'utiliser

boscoa committed:
Added translation and rotation to composite shapes

pspino reviewed (approved):
@boscoa moi je mettrais le safe guard pour l'histoire de la string null. Sinon c'est approved pour moi

boscoa committed:
Rebase onto master

pgarneau commented on src/main/java/apaintus/models/shapes/SelectionBox.java:
```diff
@@ -0,0 +1,149 @@
+        shapes = new ArrayList<>();
+    }
+
+    public void update(ShapeAttributes shapeAttributes) {
```
Quelle est l'utilite?

boscoa replied:
De?

pgarneau replied:
La methode update() avec shapeAttributes de SelectionBox

boscoa replied:
La factory l'utilise toujours.

boscoa replied:
Ah

boscoa replied:
Laisse, je comprends

pgarneau commented on src/main/java/apaintus/models/shapes/SelectionBox.java:
```diff
@@ -0,0 +1,149 @@
+                ((SelectionBox) shape).update(attribute, step);
+            }
+            updateShape(attribute, step, shape);
+            shape.update(shape.getShapeAttributes());
```
Ca sert a quoi de shape.update() quand t'as deja updateShape() qui change les valeurs necessaires?

boscoa replied:
Pour forcer l'update de la bounding box

boscoa replied:
Je pourrais qu'update la bounding box rendu là, j'avoue. Je vais update

boscoa committed:
Updated SelectionBox

Calls boundingBox update instead of update, preventing a new update on fields.
Removed update method, it was the same as DrawableShape
